File: algorithm_running/ranking_calc.py
try:
    from algorithm import evolutionary_algorithm
except Exception as e:
    print('Module does not exist.')

fun_exec_count = 0

call_count = 0  # Global variable to store the call count

def count_calls(func):
    def wrapper(*args, **kwargs):
        global call_count
        print(f"Function {func.__name__} has been called {call_count} times.")
        call_count += 1
        if call_count > 5:
            raise Exception('Budget exhausted - limit of objective function calls reached.')
        return func(*args, **kwargs)
    return wrapper

@count_calls
def objective_function(x):
    # global fun_exec_count
    # print(fun_exec_count)
    # fun_exec_count += 1
    # if fun_exec_count > 5:
    #     raise Exception('Budget exhausted - limit of objective function calls reached.')
    return x**2

def ranking_calc():
    evolutionary_algorithm(objective_function)

try:
    ranking_calc()
except Exception as e:
    print(e.args[0])

File: algorithm_running/ranking_calc.cpp
#include <iostream>
#include <stdexcept>
// #include "ranking_calc.h"
#include "algorithm.h"

int call_count = 0; // Global variable to store the call count

// Objective function
int objective_function(int x) {
    call_count++;
    if(call_count > 5) {
        throw std::runtime_error("Budget exhausted - limit of objective function calls reached.");
    }
    return x * x;
}

// Ranking calculation
void ranking_calc() {
    evolutionary_algorithm(objective_function);
}

int main() {
    try {
        ranking_calc();
    } catch (const std::runtime_error& e) {
        std::cout << e.what() << std::endl;
    }
    return 0;
}



























// Function decorator to count function calls
// ObjectiveFunction count_calls(ObjectiveFunction func) {
//     return [func](int x) {
//         std::cout << "Function has been called " << call_count << " times." << std::endl;
//         call_count++;
//         if (call_count > 5) {
//             throw std::runtime_error("Budget exhausted - limit of objective function calls reached.");
//         }
//         return func(x);
//     };
// }